File: ProblemSet/middleSchool/second/2002/2选数/2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int n, k, ans;
int a[20];
int b[20];

int cnt;

int check (int n)
{
    int i;

    for (i = 2; i < n; i++) {
        if (n % i == 0) {
            return 0;
        }
    }

    return 1;
}

int search(int z, int total, int q)
{
    int i;


    if (q == k && check(total)) {
        ans++;
        return 0;
    }

    for (i = z+1; i < n; i++) {
        if (b[i] != 1) {
            b[i] = 1;
            search(i, total+a[i], q+1);
            b[i] = 0;
        }
    }

    return 0;
}

int main (int argc, char const* argv[])
{
    int i;
    ans = 0;
    
    cnt = 0;
    memset(b, 0, sizeof(int) * 20);

    scanf("%d %d", &n, &k);

    for (i = 0; i < n; i++) {
        scanf("%d", &a[i]);
    }

    for (i = 0; i < n; i++) {
        b[i] = 1;
        search(i, a[i], 1);
        b[i] = 0;
    }

    printf("%d\n", ans);

    return 0;
}

File: ProblemSet/middleSchool/second/2002/2选数/2.pas
program p67;
var
    ans,n,i,k:longint;
    num:array[1..20] of longint;
    store:array[1..20] of boolean;
     
function check(x:longint):boolean;
var
i:longint;
begin
    if x=1 then
        exit(false);
    if x=2 then
        exit(true);
    if x=3 then
        exit(true);
    for i:=2 to trunc(sqrt(x)) do
        if (x mod i)=0 then
            exit(false);
exit(true);
end;
                                             
procedure search(z,total,q:longint);
var
    s:longint;
begin
    if q=k then
    begin
        if check(total) then
        begin
            writeln(total);
            inc(ans);
            exit;
        end;
    end;
    for s:=z+1 to n do
    begin
        if not(store[s]) then
            begin
                store[s]:=true;
                search(s,total+num[s],q+1);
                store[s]:=false;
            end;
    end;
end;
 
begin
    readln(n,k);
    for i:=1 to n do
        read(num[i]);
        readln;
           
    for i:=1 to n do
    begin
        store[i]:=true;
        search(i,num[i],1);
        store[i]:=false;
    end;
    writeln(ans);
end.
